src/txmempool.cpp

```diff
@@ -889,6 +960,140 @@ void CTxMemPool::removeForBlock(const std::vector<CTransactionRef> &vtx,
     std::list<CTransactionRef> &conflicts,
     bool fCurrentEstimate,
     std::vector<CTxChange> *txChanges)
+{
+    WRITELOCK(cs_txmempool);
+
+    // Process the block for transasction removal and ancestor state updates.
+    // As a first step remove all the txns that are in the block from the mempool by
+    // first updating the children for removal of the parent, and also saving the entries
+    // for policy estimate updates.  Then in the final loop simply remove all the txns.
+    //
+    // While we're updating the child transactions prior to removal we can also gather the
+    // mapTxnChaiTips.
```

Comment:
yes and no. They are a subset of the endpoints. Where the endpoints could be either the last transaction in the chain or the first.  The chaintip is just the first tx in the chain...however we also extend that by saying that a chaintip can also be a place where two transactions are inputs to another transaction (a convergence)...This is a critical edge case which took me forever to find a simple solution to, so by also making those convergences into chain tips we're able to handle those very complex transaction graphs that contain both "divergences" and "convergences"...or more simply put, transactions that split up into many and then combine later back into one .

Reply:
I see. 

last tx in a chain == the youngest tx in the chain? 

I mean the one that has not descendant? 

If this is the case could it mean that a chain like that could have multiple chaintips (t2 and t3)

```
   t2
t1/
  \
   t3   
```

correct?

Reply:
Yes , a chain can have multiple tips...for example if we mined tx1, we would end up with two chain tips, tx2 and tx3.

This chain is visualized vertically. tx1 was the first tx with two outputs.  tx2 and tx3 are decendants of tx1, tx4 is a decendant of both tx2 and tx3.
      tx1
     /     \
  tx2   tx3
    \     /
     tx4

Reply:
in this case how is tx1 not double counted in the statistics of tx4 in the UpdateTxnChainState call?

Reply:
I think I know what you're asking, but actually in this case tx1 doesn't factor into the equation because it was mined...tx2 and tx3 become the chaintips.  If tx1 wasn't mined though and we needed to update state for this chain, then tx4 become an additional chaintip because it has more than 1 parent and since chaintips get their state updated separately it ends up having the correct count.